### DISCUSSION TITLE
Is there a way to add custom headers to all http requests and then add them to metrics?

### DISCUSSION THREAD

**Discussion by tuhao1020** [open-telemetry/opentelemetry-java-instrumentation#7704] [Q&A] [answered]
For traces we can create a custom ResourceProvider in extensions, can we do the same for metrics? 
For example, add a new extension, add custom headers to all http requests, and then add them to metrics as dimensions.

**Comment by mateuszrzeszutek** [comment]
Hey @tuhao1020 ,

> For traces we can create a custom ResourceProvider in extensions, can we do the same for metrics?

The same resource attributes are attached to all telemetry signals; if you register a `ResourceProvider` it will apply both to traces and metrics.

> Is there a way to add custom headers to all http requests...

Like, append extra headers to outgoing HTTP client requests?
Propagators kinda do that, they allow [injecting](https://opentelemetry.io/docs/reference/specification/context/api-propagators/#inject) custom key-value pairs to outgoing message headers (that includes HTTP, RPC, messaging, etc). 

You can also use [baggage](https://opentelemetry.io/docs/reference/specification/baggage/api/) and the baggage propagator for that.

> ...and then add them to metrics as dimensions

If you're using baggage, you can append chosen baggage attributes to the metrics; that requires using [an internal/experimental API](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/SdkMeterProviderUtil.java#L55) though.

**Reply by tuhao1020** [reply]
this is the extension example code
```
@Override
  public void customize(AutoConfigurationCustomizer autoConfiguration) {
    autoConfiguration
        .addTracerProviderCustomizer(this::configureSdkTracerProvider)
        .addPropertiesSupplier(this::getDefaultProperties);
  }
```
and there is a `DemoPropagator.java` but seems not work, can you give me some more  sample code?

**Comment by tuhao1020** [comment]
@mateuszrzeszutek  can you show me how to register SdkMeterProvider to AutoConfigurationCustomizerProvider and then  use `SdkMeterProviderUtil.appendFilteredBaggageAttributes` ?

**Reply by mateuszrzeszutek** [reply, marked as answer]
Try the following:

```java
@AutoService(AutoConfigurationCustomizerProvider.class)
class SdkMeterProviderConfigurer implements AutoConfigurationCustomizerProvider {
  
  public void customize(AutoConfigurationCustomizer autoConfiguration) {
    autoConfiguration.addMeterProviderCustomizer((sdkMeterProviderBuilder, configProperties) -> {
      ViewBuilder viewBuilder = View.builder();
      SdkMeterProviderUtil.appendFilteredBaggageAttributes(viewBuilder, name -> name.equals("my-baggage-attribute"));
      View view = viewBuilder.build();
      return sdkMeterProviderBuilder.registerView(InstrumentSelector.builder().setName("http.client.duration").build(), view);
    });
  }
}
```
